src/sentry/api/endpoints/release_thresholds/health_checks/is_crash_free_rate_healthy.py

```diff
@@ -0,0 +1,139 @@
+    if environment:
+        filters.append("environment")
+    if status:
+        filters.append("session.status")
+    group_series: list[dict[str, Any]] = [
+        g.get("series", {})
+        for g in sessions_data.get("groups", [])
+        if g.get("by", {}).get("release") == release_version
+        and g.get("by", {}).get("project") == project_id
+        and (g.get("by", {}).get("environment") == environment if environment else True)
+        and (g.get("by", {}).get("session.status") == status if status else True)
+    ]
+    total = 0
+    for group in group_series:
+        series = group.get(field, [])
+        if len(series) < start_idx or len(series) < end_idx:
+            logger.warning(
+                "Error calculating session totals",
+                extra={"start_idx": start_idx, "end_idx": end_idx, "series_length": len(series)},
+            )
+            raise IndexError("Start/End indexes are out of range for series")
+        total += sum(series[start_idx : end_idx + 1])
+    return total
+
+
+def get_interval_indexes(intervals: list[str], start: datetime, end: datetime) -> Tuple[int, int]:
+    """
+    :param intervals: if timestamps from fetched sessions data
+    :param start: timestamp
+    :param end: timestamp
+
+    :return: If any subsection of the intervals fall within the given start/end, return the indexes of both the start and the end
+
+    TODO: raise if start/end not contained within intervals?
+    """
+    start_idx = len(intervals)
+    end_idx = 0
+
+    for idx, i in enumerate(intervals):
+        interval_date = parser.parse(i).replace(tzinfo=None)
+        if start <= interval_date < end:
+            if idx < start_idx:
+                start_idx = idx
+            if end_idx < idx:
+                end_idx = idx
+
+    return start_idx, end_idx
+
+
+def is_crash_free_rate_healthy_check(
+    ethreshold: EnrichedThreshold,
+    sessions_data: Dict[str, Any],
+    display: str,
+) -> Tuple[bool, float]:  # (is_healthy, metric_value)
+    """
+    Derives percent from crash total over total count
+
+    NOTE: Logic mirrors Frontend views/releases/list/releasesRequest::getCrashFreeRate
+    NOTE: if 'environment' not provided, will tally sums for _all_ sessions data as opposed to only sessions with environment of 'None'
+    """
+    field = "count_unique(user)" if display == CRASH_USERS_DISPLAY else "sum(session)"
+    release_version = ethreshold["release"]
+    project_id = ethreshold["project_id"]
+    environment = ethreshold["environment"]
+    intervals = sessions_data.get("intervals", [])
+    try:
+        start_idx, end_idx = get_interval_indexes(
+            intervals=intervals, start=ethreshold["start"], end=ethreshold["end"]
+        )
+        if start_idx > end_idx:
+            logger.warning("Threshold window not within provided session data")
+            raise ValueError("Threshold window not within provided session data")
+    except ValueError:
+        return False, 0
```

Comment:
should we return something other than `0` to indicate an error occurred rather than assuming no crashes occurred? if the data passed in is somehow faulty is it the same as there actually being 0 crashes

Reply:
That's a fair point 🤔 

Maybe -1?

Reply:
maybe, but i'm not sure if the frontend needs to show something different for a `-1`

Reply:
Yea - we can make that change

I'll update it to -1 for now